src/home_robot/home_robot/agent/ovmm_agent/llm_agent.py

```diff
@@ -0,0 +1,169 @@
+from home_robot.manipulation import HeuristicPlacePolicy
+
+
+class Skill(IntEnum):
+    NAV_TO_OBJ = auto()
+    GAZE_AT_OBJ = auto()
+    PICK = auto()
+    NAV_TO_REC = auto()
+    GAZE_AT_REC = auto()
+    PLACE = auto()
+
+
+def get_skill_as_one_hot_dict(curr_skill: Skill):
+    skill_dict = {skill.name: 0 for skill in Skill}
+    skill_dict[f"is_curr_skill_{Skill(curr_skill).name}"] = 1
+    return skill_dict
+
+
+class LLMAgent(OpenVocabManipAgent):
+    """Simple object nav agent based on a 2D semantic map."""
+
+    def __init__(self, config, device_id: int = 0, obs_spaces=None, action_spaces=None):
+        super().__init__(config, device_id=device_id)
+
+        self.device = "cuda" if torch.cuda.is_available() else "cpu"
+        self.model, self.preprocess = clip.load("ViT-B/32", device=self.device)
+        # self.goal_obs = {}
+        self.clip_threshold = 0.8
+        self.num_object_pixel_threshold = 500
+
+        self.object_of_interest = None
+        self.object_names = None
+
+        self.memory = {}
+
+    def _save_image(self, rgb):
+        """Simple helper function to save images for debug"""
+        im = Image.fromarray(rgb)
+        im.save("datadump/test.png")
+
+    def _update_memory(self, obs):
+        text = clip.tokenize(self.object_names).to(self.device)
+        # curr_pose = np.array([obs.gps[0], obs.gps[1], obs.compass[0]])
+
+        # record GT label of if the object has been found
+        if obs.task_observations["object_goal"] in obs.semantic:
+            self.memory[self.timesteps[0]]["is_found"] = True
+
+        for object_id in self.object_of_interest:
+            if object_id not in obs.semantic:
+                continue
+
+            if (
+                self._get_num_goal_pixels(obs, object_id)
+                < self.num_object_pixel_threshold
+            ):
+                continue
+
+            image_arr = self._segment_goal(obs, object_id)
+            image = (
+                self.preprocess(Image.fromarray(image_arr)).unsqueeze(0).to(self.device)
+            )
+
+            with torch.no_grad():
+                image_features = self.model.encode_image(image)
+                # text_features = self.model.encode_text(text)
+                logits_per_image, logits_per_text = self.model(image, text)
+                probs = logits_per_image.softmax(dim=-1).cpu().numpy()
+                print(probs)
+                if np.max(probs) > self.clip_threshold:
+                    # if True:
+                    self._save_image(image_arr)
+                    print(self.object_names[np.argmax(probs)])
+                    print("Label probs:", probs)
+
+                    self.memory[self.timesteps[0]]["clip_features"].append(
+                        image_features.squeeze(0).tolist()
+                    )
+
+            # self.memory.append(image_features.squeeze(0))
+
+    def _segment_goal(self, obs, object_id):
```

Comment:
not clear what this fn is doing to me

Reply:
It uses ground truth semantics to identify the edges of the objects in the current frame, and then crop it out. Rewrote to improve clarity